support/build/Makefile.rules

```diff
@@ -252,30 +252,24 @@ verbose_cmd = @printf '  %-7s %s\n' '$(1)'  '$(2)' && $(3)
 verbose_cmd_inner = printf '  %-7s %s\n' '$(1)'  '$(2)' && $(3)
 endif
 
-# Writes arguments into a build_args file
-#
-# write_build_args $target,$longargs(optional)
-ifeq ($(CONFIG_BUILD_ARGS_FILE),y)
-write_build_args = $(if $(2),$(file > $(addsuffix .build_args,$(1)),$(2))@$(addsuffix .build_args,$(1)),)
-else
-write_build_args = $(2)
-endif
-
 # Calls a command that creates an object
 #
-# build_cmd $quietlabel,$libname(optional),$target,$command,$longargs(optional)
+# build_cmd $quietlabel,$libname(optional),$target,$command
 ifeq ($(CONFIG_RECORD_BUILDTIME_TIME),y)
 define build_cmd =
-$(call verbose_cmd,$(1),$(if $(2),$(2)':' ,)$(notdir $(3)),$(TIME) $(TIMEFLAGS) -o $(addsuffix .time,$(3)) $(4) $(call write_build_args,$(3),$(5)))
+$(file > $(addsuffix .cmd,$(3)),$(4))\
```

Comment:
The following will _not work_, as the file function will not be expanded.
```suggestion
$$(file > $(addsuffix .cmd,$(3)),$(4))\
```

Reply:
hum, but it is a question now if the double `$` would even make sense here. The thing is that the variables `$(3)` and `$(4)` are derived from `$$()` variables. Probably no need to take care here.

Reply:
Did you see any problems with the second expansion?

Reply:
This is exactly the problem, the $$() variables will be written into the file unexpanded.

Reply:
Okay, sorry, I just found this: https://www.gnu.org/software/make/manual/html_node/Secondary-Expansion.html

Reply:
That will leads to outputting `$(file > $(3).cmd,$(4))`. But even if this would be expanded later, outside the context of `build_cmd`, `$(3)` has a different value then.

I think we are missing something:

- If the $$arguments would expanded before expanding `build_cmd`, `$(file` would work.
- If the $$arguments would be expanded after expanding `build_cmd` `$$(file` should work
- If the $$arguments would not be expanded a second time, the the build system would not work.

Running `make -n` I noticed $$ variables (and file) written into `build_cmd` are sometimes expanded and sometimes not. I will look into that.

Reply:
It seem like the second expansion is only happening for rules that are generated by an expansion (`buildrule_*`), but not for (non-generated) rules were `build_cmd` is called only in the recipe.

```
app-helloworld_kvm-arm64
app-helloworld_kvm-arm64.dbg
app-helloworld_kvm-arm64.gz
app-helloworld_kvm-arm64.ld.o
app-helloworld_kvm-arm64.o
provided_syscalls.h
provided_syscalls.h.in.new
syscall_map.h.new
syscall_nrs2.h.new
syscall_nrs.h.new
syscall_r_static.h
syscall_static.h
syscall_stubs.h.new
```


It makes more sense for me now. In generated build rules, the variables should be expanded only when the recipe is expanded, not when the rule itself is generated. This is probably why the $$ variables are there.
I am would suspect this is not even as secondary expansion as described in the [manual](https://www.gnu.org/software/make/manual/html_node/Secondary-Expansion.html), but the rule generation is an additional prior extension the $$ variables need to be escaped from.

Reply:
> POssible that we need to handle a path with eval differently from cases where build_cmd is directly invoked.

Yes, this is exactly what I have done in the  current version (4cea54722fea9e44bb2235383570b4b87ccd1dde).